File: remotedsh-android/core-contract/src/main/java/dsh/mobile/contract/Frames.kt
// 帧编解码（契约 §3.1–§3.3）——与 js/frames.mjs、go/frames.go 共用 vectors/frames.json
// 11B 大端帧头：ver u8 / type u8 / flags u8 / stream_id u32 / len u32；len ≤ 16373
package dsh.mobile.contract

import java.nio.ByteBuffer

object Frames {
    const val FRAME_HEADER_SIZE = 11
    const val MAX_PAYLOAD = 16373
    const val MAX_FRAME_SIZE = 16384
    const val FRAME_VER = 1

    const val TYPE_OPEN: Byte = 0x01
    const val TYPE_DATA: Byte = 0x02
    const val TYPE_FIN: Byte = 0x03
    const val TYPE_RST: Byte = 0x04
    const val TYPE_WINDOW: Byte = 0x05
    const val TYPE_PING: Byte = 0x06
    const val TYPE_PONG: Byte = 0x07
}

class FrameException(val code: String, message: String) : Exception(message)

data class Frame(
    val ver: Int,
    val type: Int,
    val flags: Int, // 接收方 MUST 忽略非 0（契约 §3.1）
    val streamId: Long,
    val payload: ByteArray,
) {
    override fun equals(other: Any?): Boolean = other is Frame &&
        ver == other.ver && type == other.type && flags == other.flags &&
        streamId == other.streamId && payload.contentEquals(other.payload)
    override fun hashCode(): Int = ver * 31 + type * 7 + streamId.hashCode()
}

object FrameCodec {
    /** 编码一帧；flags 恒写 0（发送方 MUST，契约 §3.1）。超限抛 FrameException("LEN_OVERFLOW") */
    fun encode(type: Int, streamId: Long, payload: ByteArray = ByteArray(0)): ByteArray {
        if (payload.size > Frames.MAX_PAYLOAD) {
            throw FrameException("LEN_OVERFLOW", "len=${payload.size} > ${Frames.MAX_PAYLOAD}")
        }
        if (streamId < 0 || streamId > 0xFFFFFFFFL) {
            throw FrameException("BAD_STREAM_ID", "stream_id=$streamId 超出 u32 范围（契约 §3.1）")
        }
        val out = ByteArray(Frames.FRAME_HEADER_SIZE + payload.size)
        val b = ByteBuffer.wrap(out)
        b.put(Frames.FRAME_VER.toByte()).put(type.toByte()).put(0)
        b.putInt(streamId.toInt()).putInt(payload.size)
        b.put(payload)
        return out
    }

    /** 帧头不完整抛 FrameException("HEADER_INCOMPLETE")；ver!=1 抛 "BAD_VER"；len 超限抛 "LEN_OVERFLOW" */
    fun decode(bytes: ByteArray, offset: Int = 0, length: Int = bytes.size - offset): Frame {
        if (length < Frames.FRAME_HEADER_SIZE) {
            throw FrameException("HEADER_INCOMPLETE", "only $length bytes")
        }
        val b = ByteBuffer.wrap(bytes, offset, length)
        val ver = b.get().toInt() and 0xFF
        if (ver != Frames.FRAME_VER) throw FrameException("BAD_VER", "ver=$ver")
        val type = b.get().toInt() and 0xFF
        val flags = b.get().toInt() and 0xFF
        val streamId = b.int.toLong() and 0xFFFFFFFFL
        // ByteBuffer.getInt() 返回 signed int；契约 len 是 u32（0–16373），转 unsigned 后校验
        val len = b.int.toLong() and 0xFFFFFFFFL
        if (len > Frames.MAX_PAYLOAD) throw FrameException("LEN_OVERFLOW", "len=$len")
        if (length - Frames.FRAME_HEADER_SIZE < len) throw FrameException("HEADER_INCOMPLETE", "truncated")
        val payloadLen = len.toInt()
        val payload = ByteArray(payloadLen)
        b.get(payload)
        return Frame(ver, type, flags, streamId, payload)
    }
}

File: remotedsh-contract/dart/lib/contract/frames.dart
// 帧编解码（契约 §3.1–§3.3）—— 自 remotedsh-contract/js/frames.mjs 逐字移植
// 11B 大端帧头：ver u8 / type u8 / flags u8 / stream_id u32 / len u32；len ≤ 16373（总帧 ≤ 16384）
import 'dart:convert';
import 'dart:typed_data';

const int frameHeaderSize = 11;
const int maxPayload = 16373;
const int maxFrameSize = 16384;
const int frameVer = 1;

class FrameType {
  static const int open = 0x01;
  static const int data = 0x02;
  static const int fin = 0x03;
  static const int rst = 0x04;
  static const int window = 0x05;
  static const int ping = 0x06;
  static const int pong = 0x07;

  static String name(int t) => const {
        open: 'OPEN', data: 'DATA', fin: 'FIN', rst: 'RST',
        window: 'WINDOW', ping: 'PING', pong: 'PONG',
      }[t] ?? 'UNKNOWN($t)';
}

class FrameError implements Exception {
  final String code; // 'BAD_VER' | 'LEN_OVERFLOW' | 'HEADER_INCOMPLETE'
  final String message;
  FrameError(this.code, this.message);
  @override
  String toString() => 'FrameError($code): $message';
}

class Frame {
  final int ver;
  final int type;
  final int flags;
  final int streamId;
  final Uint8List payload;
  Frame(this.ver, this.type, this.flags, this.streamId, this.payload);
}

/// 编码一帧。flags 由本库恒置 0（契约 §3.1：发送方 MUST 置 0）。
/// len > 16373 → 抛 FrameError('LEN_OVERFLOW')（发送侧防线；接收侧超限对流 RST）。
/// PING/PONG 调用方 MUST 传 streamId=0。
Uint8List encodeFrame(int type, int streamId, [List<int>? payload]) {
  final Uint8List buf;
  if (payload == null) {
    buf = Uint8List(0);
  } else if (payload is Uint8List) {
    buf = payload;
  } else {
    buf = Uint8List.fromList(payload);
  }
  if (buf.length > maxPayload) {
    throw FrameError('LEN_OVERFLOW', 'len=${buf.length} > $maxPayload');
  }
  final out = Uint8List(frameHeaderSize + buf.length);
  final dv = ByteData.sublistView(out);
  dv.setUint8(0, frameVer);
  dv.setUint8(1, type);
  dv.setUint8(2, 0); // flags 恒 0
  dv.setUint32(3, streamId, Endian.big);
  dv.setUint32(7, buf.length, Endian.big);
  out.setAll(frameHeaderSize, buf);
  return out;
}

/// 帧头不完整 → HEADER_INCOMPLETE（调用方应继续缓冲，等更多字节）
Frame decodeFrame(Uint8List bytes) {
  if (bytes.length < frameHeaderSize) {
    throw FrameError('HEADER_INCOMPLETE', 'only ${bytes.length}B');
  }
  final dv = ByteData.sublistView(bytes);
  final ver = dv.getUint8(0);
  if (ver != frameVer) throw FrameError('BAD_VER', 'ver=$ver'); // §3.1 非 1 → 立即断开
  final type = dv.getUint8(1);
  final flags = dv.getUint8(2); // 接收方 MUST 忽略非 0，不得据此拒帧
  final streamId = dv.getUint32(3, Endian.big);
  final len = dv.getUint32(7, Endian.big);
  if (len > maxPayload) throw FrameError('LEN_OVERFLOW', 'len=$len'); // §3.1 → 对流 RST
  if (bytes.length < frameHeaderSize + len) {
    throw FrameError('HEADER_INCOMPLETE', 'have ${bytes.length}, need ${frameHeaderSize + len}');
  }
  final payload = Uint8List.sublistView(bytes, frameHeaderSize, frameHeaderSize + len);
  return Frame(ver, type, flags, streamId, payload);
}

/// 逐字节流式拆帧器：跨 DataChannel 消息天然按消息分帧，但本拆帧器也容忍字节流
class FrameSplitter {
  Uint8List _buf = Uint8List(0);

  List<Frame> push(List<int> bytes) {
    final merged = Uint8List(_buf.length + bytes.length);
    merged.setAll(0, _buf);
    merged.setAll(_buf.length, bytes);
    _buf = merged;
    final out = <Frame>[];
    while (true) {
      if (_buf.length < frameHeaderSize) break;
      final dv = ByteData.sublistView(_buf);
      final len = dv.getUint32(7, Endian.big);
      if (_buf.length < frameHeaderSize + len) break;
      out.add(decodeFrame(
          Uint8List.sublistView(_buf, 0, frameHeaderSize + len)));
      _buf = Uint8List.sublistView(_buf, frameHeaderSize + len);
    }
    return out;
  }
}

/// 帧内 payload → utf8 字符串（OPEN authority 等文本负载用）
String payloadAsText(Uint8List p) => utf8.decode(p);
